### PR TITLE
fix: improved locking of assembler state

### PR DESCRIPTION
**Pull request by loshz**
## Which problem is this PR solving?

- https://github.com/honeycombio/honeycomb-network-agent/issues/209

## Short description of the changes

During the debugging of suspected load testing issues, the following bugs were found and fixed:
- Switch from `sync.Map` to regular map w/ mutex https://github.com/honeycombio/honeycomb-network-agent/commit/3b8a00882d2df96ee4dfbcf861f5ea6bdb9e2cb8
- Ensure stats accessed concurrently are explicitly atomic https://github.com/honeycombio/honeycomb-network-agent/commit/fe2ebd3ef6f48fde5b3024e8cc871fc5dcb88bed
- Remove unused `sync.Waitgroup` from TCP stream factory https://github.com/honeycombio/honeycomb-network-agent/commit/2e039ad6b095b835a69c80b4ed61d08b4ed7551e

## How to verify that this has the expected result
On my local Linux machine, I compiled the main branch and the changes in this PR, and ran them concurrently while sending several thousand HTTP requests to a random netcat server. The results are identical:

Main:
```
DBG TCP assembler stats IPdefrag=0 active_streams=0 conn_rejected_FSM=0 event_queue_length=500 goroutines=6 rejected_FSM=0 rejected_Options=0 source_dropped=0 source_if_dropped=0 source_received=4000 total_TCP_bytes=97500 total_streams=500 uptime_ms=25450
6:33PM DBG Stopping TCP assembler assembler_page_usage="pageCache: used: 0:" closed=500
```

PR:
```
DBG TCP assembler stats IPdefrag=0 active_streams=0 conn_rejected_FSM=0 event_queue_length=500 goroutines=6 rejected_FSM=0 rejected_Options=0 source_dropped=0 source_if_dropped=0 source_received=4000 total_TCP_bytes=97500 total_streams=500 uptime_ms=20920
6:33PM DBG Stopping TCP assembler assembler_page_usage="pageCache: used: 0:" closed=500
```

### CODE REVIEW

**Comment by robbkidd** [comment]
I've got a couple commits to propose tack onto these improvements:

- reuse the return vars within the function (https://github.com/honeycombio/honeycomb-network-agent/commit/b7762e26930e7337a51d5ddd0e42348bfc1baaac)
- include recording the packet count when we have a match (https://github.com/honeycombio/honeycomb-network-agent/commit/5b4cec8dbee18f0f0644943761cd231a405c29d1)
  - I think we missed adding that to both places prior to your updates in this PR. 🤭

File: assemblers/http_matcher.go [hunk 1]
```diff
@@ -58,19 +63,22 @@ func (m *httpMatcher) GetOrStoreRequest(key int64, timestamp time.Time, request
...
 // stores the Response for later lookup and returns match as nil and matchFound will be false.
 func (m *httpMatcher) GetOrStoreResponse(key int64, timestamp time.Time, response *http.Response, packetCount int) (match *entry, matchFound bool) {
-	e := &entry{
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+
+	if match, matchFound = m.messages[key]; matchFound {
+		match.response = response
+		match.responseTimestamp = timestamp
+		match.requestPacketCount = packetCount
```
Comment (robbkidd) [inline assemblers/http_matcher.go:72]: [Ope](https://github.com/honeycombio/honeycomb-network-agent/actions/runs/6708535155/job/18239139831#step:3:211). That's [on me](https://github.com/honeycombio/honeycomb-network-agent/commit/5b4cec8dbee18f0f0644943761cd231a405c29d1#diff-bc0c5acd792a72c573f4e4ac8193ce2a09bf61a05ebac5bb5d3149c218499ccdR76). 🤦🏻 

```suggestion
		match.responsePacketCount = packetCount
```

**Review by robbkidd — CHANGES_REQUESTED** [review]
I gave you bad code, @loshz. 😬

File: assemblers/http_matcher.go (hunk 1 above)
Comment (loshz) [reply assemblers/http_matcher.go:72]: Nope, it was definitely a copypasta error on my end!

Comment (robbkidd) [reply assemblers/http_matcher.go]: 🤦🏻  "request" and "response" and have become meaningless  to me at this point. 

semantic satiation

**Review by robbkidd — APPROVED** [review]

**Pull request merged**